cmd/contour/serve.go

```diff
@@ -352,15 +383,127 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
+	deposed := make(chan struct{})
+
+	// Set up the leader election
+	// Generate the event recorder to send election events to the logs.
+	// Set up the event bits
+	eventBroadcaster := record.NewBroadcaster()
+	// Broadcast election events to the config map
+	eventBroadcaster.StartRecordingToSink(&clientcorev1.EventSinkImpl{Interface: clientcorev1.New(client.CoreV1().RESTClient()).Events("")})
+	eventsScheme := runtime.NewScheme()
+	// Need an eventsScheme
+	err := coreV1.AddToScheme(eventsScheme)
+	check(err)
+	// The recorder is what will record events about the resource lock
+	recorder := eventBroadcaster.NewRecorder(eventsScheme, coreV1.EventSource{Component: "contour"})
+
+	// Figure out the resource lock ID
+	resourceLockID, isPodNameEnvSet := os.LookupEnv("POD_NAME")
+	if !isPodNameEnvSet {
+		resourceLockID = uuid.New().String()
+	}
+
+	// Generate the resource lock.
+	// TODO(youngnick) change this to a Lease object instead
+	// of the configmap once the Lease API has been GA for a full support
+	// cycle (ie nine months).
+	rl, err := resourcelock.New(
+		resourcelock.ConfigMapsResourceLock,
+		ctx.LeaderElectionConfig.Namespace,
+		ctx.LeaderElectionConfig.Name,
+		client.CoreV1(),
+		coordinationClient,
+		resourcelock.ResourceLockConfig{
+			Identity:      resourceLockID,
+			EventRecorder: recorder,
+		},
+	)
+	check(err)
+
+	// Make the leader elector, ready to be used in the Workgroup.
+	le, err := leaderelection.NewLeaderElector(leaderelection.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: ctx.LeaderElectionConfig.LeaseDuration,
+		RenewDeadline: ctx.LeaderElectionConfig.RenewDeadline,
+		RetryPeriod:   ctx.LeaderElectionConfig.RetryPeriod,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				log.WithFields(logrus.Fields{
+					"lock":     rl.Describe(),
+					"identity": rl.Identity(),
+				}).Info("elected leader")
+				close(leaderOK)
+			},
+			OnStoppedLeading: func() {
+				// The context being canceled will trigger a handler that will
+				// deal with being deposed.
+				close(deposed)
+			},
+		},
+	})
+	check(err)
+	// AddContext will generate its own context.Context and pass it in,
+	// managing the close process when the other goroutines are closed.
+	g.AddContext(func(electionCtx context.Context) {
+		log := log.WithField("context", "leaderelect")
+		if !ctx.EnableLeaderElection {
+			log.Info("Leader election disabled")
+			// if leader election is disabled, signal the gRPC goroutine
+			// to start serving and finsh up this context.
+			// The Workgroup will handle leaving this running until everything
+			// else closes down.
+			close(leaderOK)
+			<-electionCtx.Done()
+		}
+
+		log.WithFields(logrus.Fields{
+			"configmapname":      ctx.LeaderElectionConfig.Name,
+			"configmapnamespace": ctx.LeaderElectionConfig.Namespace,
+		}).Info("Started leader election")
+
+		le.Run(electionCtx)
+		log.Info("Finishing leader election")
```

Comment:
that becomes `stopped` which is in keeping with the rest of the members of this group.